File: src/com/brianc/html/HTMLParser.java
package com.brianc.html;

import java.util.AbstractMap;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;

import com.brianc.AbstractParser;
import com.brianc.dom.Node;

public class HTMLParser extends AbstractParser {
	public HTMLParser(String input) {
		super(input);
	}

	public Node parse() {
		List<Node> nodes = parseNodes();

		if (nodes.size() == 1) {
			return nodes.get(0);
		}

		return Node.createElement("html", new HashMap<>(), nodes);
	}

	private String parseTagName() {
		return consumeWhile(Character::isLetterOrDigit);
	}

	private Node parseNode() {
		switch (nextChar()) {
		case '<':
			return parseElement();
		default:
			return parseText();
		}
	}

	private Node parseText() {
		return Node.createText(consumeWhile(c -> c != '<'));
	}

	private Node parseElement() {
		assertInput(consumeChar() == '<');
		String tagName = parseTagName();
		Map<String, String> attrs = parseAttributes();
		assertInput(consumeChar() == '>');

		List<Node> children = parseNodes();

		assertInput(consumeChar() == '<');
		assertInput(consumeChar() == '/');
		assertInput(parseTagName().equals(tagName));
		assertInput(consumeChar() == '>');

		return Node.createElement(tagName, attrs, children);
	}

	private List<Node> parseNodes() {
		List<Node> nodes = new ArrayList<>();

		for (;;) {
			consumeWhitespace();

			if (eof() || startsWith("</")) {
				break;
			}

			nodes.add(parseNode());
		}

		return nodes;
	}

	private Map<String, String> parseAttributes() {
		Map<String, String> attributes = new HashMap<>();

		for (;;) {
			consumeWhitespace();

			if (nextChar() == '>') {
				break;
			}

			Entry<String, String> attribute = parseAttribute();
			attributes.put(attribute.getKey(), attribute.getValue());
		}

		return attributes;
	}

	private Entry<String, String> parseAttribute() {
		String name = parseTagName();
		assertInput(consumeChar() == '=');
		String value = parseAttributeValue();

		return new AbstractMap.SimpleEntry<>(name, value);
	}

	private String parseAttributeValue() {
		char openQuote = consumeChar();
		assertInput(openQuote == '"' || openQuote == '\'');
		String value = consumeWhile(c -> c != openQuote);
		assertInput(consumeChar() == openQuote);

		return value;
	}

	private void assertInput(boolean cond) {
		if (!cond) {
			throw new IllegalStateException();
		}
	}

	public static void main(String[] args) {
		String toParse = "<html><body><h1>Title</h1><div id=\"main\" class=\"test\"><p>Hello <em>world</em>!</p></div></body></html>";
		Node root = new HTMLParser(toParse).parse();

		System.out.println(root);
	}
}

File: src/com/brianc/html/HTMLParser.kt
package com.brianc.html

import java.util.ArrayList
import java.util.HashMap

import com.brianc.AbstractParser
import com.brianc.dom.Node

class HTMLParser(input: String) : AbstractParser(input) {

    fun parse(): Node {
        val nodes = parseNodes()

        if (nodes.size == 1) {
            return nodes[0]
        }

        return Node.createElement("html", HashMap<String, String>(), nodes)
    }

    private fun parseTagName(): String {
        return consumeWhile({ Character.isLetterOrDigit(it) })
    }

    private fun parseNode(): Node {
        when (nextChar()) {
            '<' -> return parseElement()
            else -> return parseText()
        }
    }

    private fun parseText(): Node {
        return Node.createText(consumeWhile { c -> c !== '<' })
    }

    private fun parseElement(): Node {
        assertInput(consumeChar() == '<')
        val tagName = parseTagName()
        val attrs = parseAttributes()
        assertInput(consumeChar() == '>')

        val children = parseNodes()

        assertInput(consumeChar() == '<')
        assertInput(consumeChar() == '/')
        assertInput(parseTagName() == tagName)
        assertInput(consumeChar() == '>')

        return Node.createElement(tagName, attrs, children)
    }

    private fun parseNodes(): List<Node> {
        val nodes = ArrayList<Node>()

        while (true) {
            consumeWhitespace()

            if (eof() || startsWith("</")) {
                break
            }

            nodes.add(parseNode())
        }

        return nodes
    }

    private fun parseAttributes(): Map<String, String> {
        val attributes = HashMap<String, String>()

        while (true) {
            consumeWhitespace()

            if (nextChar() == '>') {
                break
            }

            val attribute = parseAttribute()
            attributes.put(attribute.first, attribute.second)
        }

        return attributes
    }

    private fun parseAttribute(): Pair<String, String> {
        val name = parseTagName()
        assertInput(consumeChar() == '=')
        val value = parseAttributeValue()

        return Pair(name, value)
    }

    private fun parseAttributeValue(): String {
        val openQuote = consumeChar()
        assertInput(openQuote == '"' || openQuote == '\'')
        val value = consumeWhile { c -> c !== openQuote }
        assertInput(consumeChar() == openQuote)

        return value
    }

    private fun assertInput(cond: Boolean) {
        if (!cond) {
            throw IllegalStateException()
        }
    }

    companion object {

        @JvmStatic fun main(args: Array<String>) {
            val toParse = "<html><body><h1>Title</h1><div id=\"main\" class=\"test\"><p>Hello <em>world</em>!</p></div></body></html>"
            val root = HTMLParser(toParse).parse()

            println(root)
        }
    }
}
